### DISCUSSION TITLE
How are modulators displayed

### DISCUSSION THREAD

**Discussion by audio2000** [FluidSynth/fluidsynth#1383] [Q&A] [answered]
I am still trying to understand how to successfully add modulators to a soundfont for parameters such as volume envelope release in Polyphone 2.2.0 (using Linux so no access at the moment to newest release).  I have seen many people asking about this but have not found any answers how to achieve this, also no websites / tutorials found about this either.  So I currently have no way to test this.
Currently the soundfonts I saved with modulators do not seem to show signs of modulator links.

I was wondering, if modulators (linked to MIDI CC) are successfully saved inside a soundfont, how does Fluidsynth actually handle them?  Meaning where do the MIDI CC numbers/labels actually show up in an audio application that uses the Fluidsynth API?
Unless this is application specific? if so even just some basic examples of how they are generally displayed would be helpful.

I was thinking that if a soundfont contained for example CC70 linked to volume envelope release, that when a Fluidsynth based player loads this soundfont, then Fluidsynth now shows CC70 in the same list of CCs provided by Fluidsynth, for example 
CC7 Volume....
CC68 Legato
CC70 Volume Release (something like this)

or are they just linked to the MIDI CC number with no reference of the link / label anywhere?

Thank you for any info shared

**Comment by derselbst** [comment, marked as answer]
You seem to be fundamentally misunderstanding something. Modulators are part of the internal state of the synthesizer, they manipulate certain aspects of the audio, i.e. you can _hear_ them. But they do not "show up" anywhere.

So, if you define a modulator linked to e.g. CC70 controlling the volume env release, load the soundfont into fluidsynth, send CC70 events and perceive how the release env of that particular instrument - where you have defined the modulator -  changes.

**Comment by audio2000** [comment]
@derselbst 
Thank you for answering this question and for all the help you have provided, as well as for Fluidsynth itself :)
Okay, I believe that is what said at the bottom of my post after the "or".
I just was not sure if there was maybe some kind of visual indicator / CC## label that would get detected by Fluidsynth when a soundfont contains them.
I know this is not the same thing, but just using as an example similar to how Fluidsynth presents audio applications CC68=Legato and then that audio application can (if they decide to) show CC68=Legato visually so user knows it is available from Fluidsynth plugin.  Again I know that it is not the same, not on the same level.
Better example is how some SFZ players function, where if a parameter is set to a MIDI CC value in the SFZ file, then that parameter knob gets detected and shows up in the SFZ player as a visual knob/fader.

Okay, so it was my second assumption that these modulator CC links do not visually show up in Fluidsynth based plugin/application, DAW, etc.  They are just included in the soundfont and if the MIDI CC is adjusted it adjusts in that soundfont.
Just strange because then user would have no way of knowing that specific modulator linked to a CC exists within a soundfont, unless they opened up a soundfont editor such as Polyphone to check.

Alright well that means that the soundfont I saved in Polyphone with a modulator was not done correctly because the CC did not adjust the parameter that was set.  The Polyphone manual as far as I can see does not really cover this topic in great detail.

Anyways, now at least I know how the modulator should perform.
Thank you again for all your help

**Reply by derselbst** [reply]
> then that audio application can (if they decide to) show CC68=Legato visually so user knows it is available from Fluidsynth plugin. Again I know that it is not the same, not on the same level.

Exactly.

> Just strange because then user would have no way of knowing that specific modulator linked to a CC exists within a soundfont, unless they opened up a soundfont editor such as Polyphone to check.

Ideally, the idea behind those modulators is documented by the Soundfont author. And in any case, modulators are a customization, i.e. if the Soundfont sets up modulators, a MIDI file must also follow this convention. So you'll always have to combine both worlds.

I cannot comment on SFZ, sry.

You're welcome.
